internal/auth/auth.go

```diff
@@ -0,0 +1,150 @@
+package auth
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"github.com/iurnickita/gophermart/internal/store"
+	"github.com/iurnickita/gophermart/internal/token"
+)
+
+type Auth interface {
+	Register(w http.ResponseWriter, r *http.Request)
+	Login(w http.ResponseWriter, r *http.Request)
+	Middleware(h http.HandlerFunc) http.HandlerFunc
+}
+
+const (
+	HeaderUserCodeKey = "userCode"
+	cookieUserToken   = "gophermartUserToken"
+)
+
+type auth struct {
+	store store.Store
+}
+
+func NewAuth(store store.Store) (Auth, error) {
+	return &auth{store: store}, nil
+}
+
+type RegisterJSONRequest struct {
+	Login    string `json:"login"`
+	Password string `json:"password"`
+}
+
+func (a *auth) Register(w http.ResponseWriter, r *http.Request) {
+	// Чтение логина/пароля
+	var buffer bytes.Buffer
+	_, err := buffer.ReadFrom(r.Body)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	var register RegisterJSONRequest
+	err = json.Unmarshal(buffer.Bytes(), &register)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	// Запись в БД
+	ctx := context.Background()
```

Comment:
```suggestion
	ctx := r.Context()
```
Контекст в данном случае следует брать не из воздуха, а из запроса. Если пользователь (или какая-нибудь внешняя злая сила) сбросит соединение, то контекст протухнет, и запись в базу благополучно оборвется.

Аналогично далее везде. Существует одно исключение - если хендлер запускает отложенную процедуру, которая должна некоторое время жить после того, как запрос завершится. Во всех остальных случаях контекст берется из запроса.